op-program/client/interop/interop.go

```diff
@@ -40,69 +45,101 @@ func RunInteropProgram(logger log.Logger, bootInfo *boot.BootInfoInterop, l1Prei
 func runInteropProgram(logger log.Logger, bootInfo *boot.BootInfoInterop, l1PreimageOracle l1.Oracle, l2PreimageOracle l2.Oracle, validateClaim bool, tasks taskExecutor) error {
 	logger.Info("Interop Program Bootstrapped", "bootInfo", bootInfo)
 
+	expected, err := stateTransition(logger, bootInfo, l1PreimageOracle, l2PreimageOracle, tasks)
+	if err != nil {
+		return err
+	}
+	if !validateClaim {
+		return nil
+	}
+	return claim.ValidateClaim(logger, eth.Bytes32(bootInfo.Claim), eth.Bytes32(expected))
+}
+
+func stateTransition(logger log.Logger, bootInfo *boot.BootInfoInterop, l1PreimageOracle l1.Oracle, l2PreimageOracle l2.Oracle, tasks taskExecutor) (common.Hash, error) {
+	if bootInfo.AgreedPrestate == InvalidTransitionHash {
+		return InvalidTransitionHash, nil
+	}
+	transitionState, superRoot, err := parseAgreedState(bootInfo, l2PreimageOracle)
+	if err != nil {
+		return common.Hash{}, err
+	}
+	expectedPendingProgress := transitionState.PendingProgress
+	if transitionState.Step < uint64(len(superRoot.Chains)) {
+		block, err := deriveOptimisticBlock(logger, bootInfo, l1PreimageOracle, l2PreimageOracle, superRoot, transitionState, tasks)
+		if errors.Is(err, ErrL1HeadReached) {
+			return InvalidTransitionHash, nil
+		} else if err != nil {
+			return common.Hash{}, err
+		}
+		expectedPendingProgress = append(expectedPendingProgress, block)
+	}
+	finalState := &types.TransitionState{
+		SuperRoot:       transitionState.SuperRoot,
+		PendingProgress: expectedPendingProgress,
+		Step:            transitionState.Step + 1,
+	}
+	expected, err := finalState.Hash()
+	if err != nil {
+		return common.Hash{}, err
+	}
+	return expected, nil
+}
+
+func parseAgreedState(bootInfo *boot.BootInfoInterop, l2PreimageOracle l2.Oracle) (*types.TransitionState, *eth.SuperV1, error) {
 	// For the first step in a timestamp, we would get a SuperRoot as the agreed claim - TransitionStateByRoot will
 	// automatically convert it to a TransitionState with Step: 0.
 	transitionState := l2PreimageOracle.TransitionStateByRoot(bootInfo.AgreedPrestate)
 	if transitionState.Version() != types.IntermediateTransitionVersion {
-		return fmt.Errorf("%w: %v", ErrIncorrectOutputRootType, transitionState.Version())
+		return nil, nil, fmt.Errorf("%w: %v", ErrIncorrectOutputRootType, transitionState.Version())
```

Comment:
I missed this before. But how does this interact with trace extension?
I understand that if the agreed state is a fully derived super root, then by construction, the root must have been derived before reaching its L1Head. In which case, shouldn't we fill in the missing states rather than return an error.

Reply:
clarified offline. The super root version is also an `IntermediateTransitionVersion`. So this is OK.